👋 **Приветствую, Михаил!**

Меня зовут Арсен Абдулин, и я буду твоим ревьювером по проекту. Предлагаю общаться на «ты», если не против =) Моя цель — не поиск ошибок в твоей работе, а поделиться опытом и помочь успешно завершить проект, чтобы на шаг приблизиться к профессии специалиста по Data Science.

Я вижу, что ты проделал большую работу, она выполнена на хорошем уровне 😊 В целом ты получил положительный результат по проекту, но есть немного рекомендаций, я отметил их в комментариях! 

При проверке работ я делаю следующие комментарии:

<div class="alert alert-success">
<b>✔️ Зеленым цветом</b> отмечены удачные решения, на которые можно опираться в следующих проектах.</div>

<div class="alert alert-warning">
<b>⚠️ Желтым цветом</b> я отметил пункты, которые в следующий раз можно сделать по-другому. Одно-два таких замечания в проекте допускается, но если их много — проект следует доработать. </div>

<div class="alert alert-danger">
<b>🚫 Красным цветом</b> отмечены критические замечания, которые необходимо поправить, чтобы принять проект. </div>

Если какие-то моменты в задании для тебя были непонятны и у тебя есть ко мне вопросы — смело спрашивай 😊 Также ты можешь доработать места, где есть желтые комментарии в этом проекте (однако, это не обязательно).

Предлагаю работать над проектом в диалоге: если ты решишь что-то поменять по моим рекомендациям — пиши об этом (выбери для своих комментариев определенный цвет - так мне будет легче увидеть изменения). Пожалуйста не перемещай, не изменяй и не удаляй мои комментарии. Все это поможет сделать проверку твоего проекта оперативнее.

<div class="alert alert-info"> <b>ℹ️ Комментарий студента: </b> Пример комментария. </div>

<font color=blue>
<b> Вступление:</b> Работа выполнена на хорошем уровне! Ввиду необъятности тематики по анализу текстов, хочу предложить несколько полезных ссылок от ревьюверов (если заинтересует):
    
Основы Natural Language Processing для текста  — https://habr.com/ru/company/Voximplant/blog/446738/
    
Интересная статья в общем по NLP — https://habr.com/ru/company/oleg-bunin/blog/352614/
    
BERT, ELMO и Ко в картинках (как в NLP пришло трансферное обучение) — https://habr.com/ru/post/487358/
    
В частности, BERT является довольно популярной моделью. Вот еще информация по этой модели:
    
https://habr.com/ru/post/436878/
    
http://jalammar.github.io/illustrated-bert/; https://huggingface.co/docs/transformers/model_doc/bert
    
И другие ссылки: 
    
http://jalammar.github.io/illustrated-word2vec/
    
https://colah.github.io/posts/2015-08-Understanding-LSTMs/
    
http://jalammar.github.io/illustrated-transformer/
    
https://colab.research.google.com/drive/1SBha91u6142pW5mSiQOOjJWCCSAFj4HI?usp=sharing    
 
</font>   

Может быть тебе будет интересно ознакомиться с примером BERT с GPU:

```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)</font>

<div class="alert alert-info"> <b>ℹ️ Комментарий студента: </b> <br>

**Привет, Арсен!** <br>
    
Огромное спасибо за такое большое количество дополнительных материалов и комментариев, кое-что я уже открыл почитать! <br>

Ещё в тренажёре я обратил внимание на некоторые вещи, например на использование progress bar, которые хотел бы попробовать. В ходе работы (это был третий вариант после 2-х тупиков с последующими откатами версии) я обратил внимание на то, что модели с типично хорошими результами (RandomForest и LGBM) для текстов дают не такие хорошие результаты, как в прошлых проектах из чего я сделал вывод, что для работы с текстами следует применять специальные библиотеки типа `Bert`, но так и не предпринял попыток в этом направлении по 3-м причинам: <br>
    - мои домашние дела, к сожалению, не дают мне достаточной свободы по времени для углубления в эту область на должном уровне; <br>
    - мой склад ума не позволяет мне сходу вникать во все детали из-за чего я большинство материалов пересматриваю и перечитываю по 2-3 раза, т.к. после первого восприятия у меня появляется только навык, но не понимание. Для меня только глубокое понимание фундаментальных основ позволяет закрепить полученные знания, т.е. пока я не разберусь в математической сути методов, пока не увижу красоту разработанных кем-то подходов, для меня это останется комбинацей последовательности действий, которую я в лучшем случае смогу просто повторить. Поэтому, если в будущем дойдёт до этого, то я буду ещё не один раз перечитывать и пробовать всё, что было в тренажёре и всё что ты прислал; <br>
    - Работа с текстами показалась мне интересной, но при этом специфичной. По основному своему опыту работы я - инженер-технолог с 10-ти летним стажем и всё, что было на курсе до этого я уже мысленно "прикрутил" к своей работе и у меня уже, как в мультике про Скруджа МакДака, в глазах крутятся цифры с сэкономленными деньгами от внедрения ML в разных производственных отделах. Но в производcтве нет работы с текстами и в этом смысле текущий модуль я рассматриваю со стороны спортивного интереса, по крайней мере пока. Однако, кто знает, может быть в будущем я уйду с головой в IT и DS, и тогда, возможно, я займусь вопросом обработки текстов по полной. <br>
    
В общем спасибо ещё раз за материалы и комментарии. Вопросов по этой части работы у меня нет. По дополнительным материалам и моделям они, возможно, появятся позже, но сперва я всё это перечитаю и попробую сам. <br>
    
До новых встреч!
</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера (v2):</b> Михаил, спасибо за такой развернутый комментарий! Мне тоже есть чем поделиться и заодно отвечу на твои вопросы =)
    
По использованию progress bar можно посмотреть здесь: https://habr.com/ru/post/483400/
    
Модели деревьев (RandomForest, DecisionTree) в задачах NLP действительно работают не очень хорошо. LGBM, CatBoost показывают неплохие результаты. LogisticRegression быстро обучается и приемлемое качество.
    
Чтобы улучшить результаты, векторизацию текста делают с помощью нейросетей. Полученные таким образом вектора (эмбеддинги) можно передать в классическую модель, ту же LogisticRegression, результаты будут гораздо выше (> 0.9).
    
Перечитывать материалы по нескольку раз это нормально, информации действительно много. Вообще я для себя всегда делаю конспект, так легче запоминается. К сожалению, когда времени мало, не всегда удается добраться до сути, поэтому в каких-то вещах стоит все принять как есть, а понимание придет позже =)
    
Я прекрасно тебя понимаю, т.к. сам инженер по образованию, в DS пришел только в 2020 году, до этого около 10 лет занимался проектированием и расчетами. Также всегда старался вникнуть в суть дела, подолгу разбирался, это конечно приносило пользу. Но когда окончил свой первый курс по DS понял что информации очень много и понимание начало приходить, только когда стал решать самостоятельные задачи.
    
Работу инженером я завершил, сейчас работаю и развиваюсь в области DS.
    
Желаю тебе успешно закончить курс, найти свое место в профессии Data Science, и обязательно проси повышение на работе, если там планируешь применять ML =)
</div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Знакомство-с-данными" data-toc-modified-id="Знакомство-с-данными-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Знакомство с данными</a></span></li><li><span><a href="#Предобработка" data-toc-modified-id="Предобработка-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Предобработка</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#LogisticRegression" data-toc-modified-id="LogisticRegression-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>LogisticRegression</a></span></li><li><span><a href="#RandomForestClassifier" data-toc-modified-id="RandomForestClassifier-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>RandomForestClassifier</a></span></li><li><span><a href="#LGBMClassifier" data-toc-modified-id="LGBMClassifier-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>LGBMClassifier</a></span></li><li><span><a href="#LogisticRegression---CountVectorizer" data-toc-modified-id="LogisticRegression---CountVectorizer-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>LogisticRegression - CountVectorizer</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Проверка-моделей-на-тестовой-выборке" data-toc-modified-id="Проверка-моделей-на-тестовой-выборке-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Проверка моделей на тестовой выборке</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Хорошо, что добавил небольшое описание и содержание проекта.

</div>

## Подготовка данных

### Импорт библиотек

In [1]:
# Импорт библиотек.
import matplotlib.pyplot as plt
import nltk
import numpy as np
import pandas as pd
import re
import seaborn as sns
import sys
import warnings

# Импорт модулей IPython.
from IPython.display import display_html
from IPython.display import display, HTML, Math, Latex

# Импорт метрик, векторизаторов, лемматизатора и CV
from nltk.stem import WordNetLemmatizer 
from nltk.corpus import stopwords as nltk_stopwords
from sklearn.feature_extraction.text import TfidfVectorizer,CountVectorizer 
from sklearn.metrics import make_scorer,accuracy_score,f1_score,precision_score,recall_score
from sklearn.metrics import precision_recall_curve,roc_auc_score,roc_curve
from sklearn.model_selection import GridSearchCV,cross_val_score,train_test_split
from sklearn.pipeline import Pipeline
from sklearn.utils import shuffle

# Импорт моделей.
from lightgbm import LGBMClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

# Импорт динамически обновляемого индикатора выполнения
from tqdm import notebook


nltk.download('popular')
stopwords = set(nltk_stopwords.words('english'))

pd.set_option('display.max_columns', None)         # Снимем ограничение на количество показываемых столбцов
pd.options.mode.chained_assignment = None          # Скрываем предупреждение при масшатбировании
pd.options.display.max_rows = None                 # Сброс ограничений на количество выводимых рядов
pd.set_option('display.max_colwidth', None)        # Сброс ограничений на количество символов в записи
warnings.filterwarnings('ignore')                  # Игнорируем предупреждения

[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to
[nltk_data]    |     C:\Users\Mikhail\AppData\Roaming\nltk_data...
[nltk_data]    |   Package cmudict is already up-to-date!
[nltk_data]    | Downloading package gazetteers to
[nltk_data]    |     C:\Users\Mikhail\AppData\Roaming\nltk_data...
[nltk_data]    |   Package gazetteers is already up-to-date!
[nltk_data]    | Downloading package genesis to
[nltk_data]    |     C:\Users\Mikhail\AppData\Roaming\nltk_data...
[nltk_data]    |   Package genesis is already up-to-date!
[nltk_data]    | Downloading package gutenberg to
[nltk_data]    |     C:\Users\Mikhail\AppData\Roaming\nltk_data...
[nltk_data]    |   Package gutenberg is already up-to-date!
[nltk_data]    | Downloading package inaugural to
[nltk_data]    |     C:\Users\Mikhail\AppData\Roaming\nltk_data...
[nltk_data]    |   Package inaugural is already up-to-date!
[nltk_data]    | Downloading package movie_reviews to
[nlt

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Необходимые библиотеки импортированы.

</div>

### Загрузка данных

In [2]:
try:
    toxic_comments = pd.read_csv('/datasets/toxic_comments.csv',sep=',')
except:
    toxic_comments = pd.read_csv('toxic_comments.csv',sep=',')

### Знакомство с данными

In [3]:
def data_info (df):
    '''
    Сводка по данным
    '''
    display(df.head(10))
    print('Общая информация:')
    display(df.info())
    print('Общая статистика:')
    display(df.describe())
    print('Дубликаты:')
    display(df.duplicated().sum())
    display(df[df.duplicated()])

In [4]:
data_info(toxic_comments)

text  \
0                                                                                                                                                                                                                                                                                                                                                                           Explanation\nWhy the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I voted at New York Dolls FAC. And please don't remove the template from the talk page since I'm retired now.89.205.38.27   
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    D'aww! He matches this background colour I'm seemingly stuck with. Thanks.  (talk) 21:51, January 11, 2016 (UTC)   
2                                                                                                                                                                                                                                                                                                                                                                                                           Hey man, I'm really not trying to edit war. It's just that this guy is constantly removing relevant information and talking to me through edits instead of my talk page. He seems to care more about the formatting than the actual info.   
3  "\nMore\nI can't make any real suggestions on improvement - I wondered if the section statistics should be later on, or a subsection of ""types of accidents""  -I think the references may need tidying so that they are all in the exact same format ie date format etc. I can do that later on, if no-one else does first - if you have any preferences for formatting style on references or want to do it yourself please let me know.\n\nThere appears to be a backlog on articles for review so I guess there may be a delay until a reviewer turns up. It's listed in the relevant form eg Wikipedia:Good_article_nominations#Transport  "   
4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 You, sir, are my hero. Any chance you remember what page that's on?   
5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 "\n\nCongratulations from me as well, use the tools well.  · talk "   
6                                                                                                                                                                                                 

Общая информация:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159571 non-null  object
 1   toxic   159571 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


None

Общая статистика:


toxic
count  159571.000000
mean        0.101679
std         0.302226
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         1.000000

Дубликаты:


0

Empty DataFrame
Columns: [text, toxic]
Index: []

***Выводы:*** <br>

- *Размер выборки 159571 строка*
- *Пропуски Отсутствуют*
- *Дубликаты Отсутствуют*
- *Присутствуют символы переноса строки и знаки препинания, т.е. текст нуждается в очистке*
- *Некоторые фразы написаны CAPSLOCK'ом*

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Здорово, провел анализ данных. 
    
Видно, что есть дисбаланс классов, это нужно будет учитывать в моделях.

</div>

###  Предобработка

In [5]:
def clearing_text(df,column_features):
    
    '''
    Функция проводит очистку анализируемого текста и создает корпус
    '''
    
    text_corp = df[column_features]
    text_corp = text_corp.apply(lambda x: " ".join((re.sub(r'[^a-zA-Z ]', ' ', x)).split()))
    
    return text_corp



def lemmatize(text):
    
    '''
    Функция проводит лемматизацию текста
    '''
    
    lemmatizer = WordNetLemmatizer()
    word_lem = corpus.apply(lambda x: [lemmatizer.lemmatize(w,'n') for w in nltk.word_tokenize(x)])
    lemmatized = word_lem.apply(lambda text:' '.join(text))
    
    return lemmatized



def lower_case(text):
    
    '''
    Функция приводит текст к нижнему регистру
    '''
    
    text_lower = text.map(lambda x: x if type(x)!=str else x.lower())
    
    return text_lower

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Лемматизация и очистка текста выполнены верно, код оформлен аккуратно =)

</div>

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> На будущее, если будет интересно более углубленное изучение анализа текстов, можешь попробовать другие способы лемматизации. 
    
1. Лемматизация с помощью Spacy. Но если делать на всем датасете, может занять 0,5-1 час, в зависимости от возможностей компьютера.
https://spacy.io/ или https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/

2. Использовать POS-теги, это сделает лемматизацию точнее:
https://webdevblog.ru/podhody-lemmatizacii-s-primerami-v-python/
    
Это занимает больше времени, поэтому можешь проверить локально, на небольшой выборке. Также используют метод pos_tag_sents(), когда перебор идет не по словам, а по предложениям — это немного быстрее.
</div>

In [6]:
corpus = clearing_text(toxic_comments,'text')
corpus = lemmatize(corpus)
corpus = lower_case(corpus)

In [7]:
corpus.head(10)

0                                                                                                                                                                                                                                                                                                                                                           explanation why the edits made under my username hardcore metallica fan were reverted they weren t vandalism just closure on some gas after i voted at new york dolls fac and please don t remove the template from the talk page since i m retired now
1                                                                                                                                                                                                                                                                                                                                                                                                                   

***Вывод:*** <br>

- *Текст комментриев был очищен он переносов и знаков препинания*
- *Была проведена лемматизация текста*
- *Весь текст был переведён в нижний регистр*

In [8]:
def split_sample(text, target):
    
    '''
    Функция делит данные на обучающую и тестовую выборки
    '''
    
    features_train, features_test, target_train, target_test = train_test_split(text, target, train_size=0.80,random_state=12345)
    
    return (features_train, features_test, target_train, target_test)

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Можно добавить параметр stratify = target, это обеспечит баланс классов в выборках.
</div>

In [9]:
toxic = toxic_comments['toxic']

features_train, features_test, target_train, target_test = split_sample(corpus, toxic)

print('Размер обучающей выборки:',features_train.shape, target_train.shape)
print('Размер тестовой выборки:',features_test.shape, target_test.shape)

Размер обучающей выборки: (127656,) (127656,)
Размер тестовой выборки: (31915,) (31915,)


In [10]:
def vectorization(train_text, test_text, vectorizer):
    
    '''
    Функция проводит векторизацию текста
    '''
    
    counter = vectorizer(stop_words=stopwords)
    vectorized_train = counter.fit_transform(train_text)
    vectoruzed_test = counter.transform(test_text)
    
    
    return vectorized_train, vectoruzed_test

In [11]:
tf_idf_train,tf_idf_test = vectorization(features_train,features_test,TfidfVectorizer)

In [12]:
print('Размер матрицы TF-IDF для обучающей выборки:',tf_idf_train.shape)
print('Размер матрицы TF-IDF для тестовой выборки:',tf_idf_test.shape)

Размер матрицы TF-IDF для обучающей выборки: (127656, 144110)
Размер матрицы TF-IDF для тестовой выборки: (31915, 144110)


***Вывод:*** <br>

- *корпус текстов разделён на обучающую и тестовую выборки*
- *была проведена векторизация текстов каждой из выделенных выборок*

## Обучение

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Можешь также попробовть Pipeline + GridSearch. Использование Pipeline позволяет избежатки утечки данных при кросс-валидации.
    
https://medium.com/analytics-vidhya/ml-pipelines-using-scikit-learn-and-gridsearchcv-fe605a7f9e05
    
https://habr.com/ru/post/538458/
    
Pipeline + GridSearchCV: https://towardsdatascience.com/nlp-with-pipeline-gridsearch-5922266e82f4
    
Таким образом в GridSearchCV можно добавить пайплайн — но это не обязательно для выполнения данной работы.
</div>

In [13]:
def grid_search_params (model,params,features,target):
    
    '''
    Принимает на вход название обучаемой модели, набор её параметров, обучающую и тестовую выборки
    Возвращает наилучшие параметры модели при которых целевая метрика f1 (среднее гармоническое полноты и точности) максимальна
    '''
    
    model_grid = GridSearchCV(model,params,cv=3, n_jobs=-1,scoring='f1', verbose=2)
    model_grid.fit(features,target)
    best_params = model_grid.best_params_
    best_score = model_grid.best_score_
    
    return model_grid,best_params,best_score

In [14]:
models_scores_list = {}

### LogisticRegression

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Здорово, что делаешь перебор параметров в моделях. Для логистической регрессии можно поварьировать константу регуляризации "C", она хорошо влияет на метрику.
</div>

In [15]:
params_lr = {"max_iter":range(63,67,1)}
model_lr = LogisticRegression(random_state = 12345,class_weight='balanced',n_jobs = -1, max_iter=250)

In [16]:
%%time
model_grid_lr,best_params_lr,best_score_lr = grid_search_params(model_lr,params_lr,
                                                                         tf_idf_train,target_train)

display(HTML(f"<font color='green'>f1 LogisticRegression: {best_score_lr:.3f}</font>"))
best_params_lr

Fitting 3 folds for each of 4 candidates, totalling 12 fits


CPU times: total: 453 ms
Wall time: 27.2 s


{'max_iter': 65}

In [17]:
models_scores_list['LogisticRegression'] = [best_params_lr,best_score_lr]

### RandomForestClassifier

In [18]:
params_forest = {'max_depth': range(398,401,1),'n_estimators': range(8,10,1)}
model_forest = RandomForestClassifier(random_state = 12345,class_weight='balanced',n_jobs = -1)

In [19]:
%%time
model_grid_forest,best_params_forest,best_score_forest = grid_search_params(model_forest,params_forest,
                                                                         tf_idf_train,target_train)

display(HTML(f"<font color='blue'>f1 RandomForestClassifier: {best_score_forest:.3f}</font>"))
best_params_forest

Fitting 3 folds for each of 6 candidates, totalling 18 fits


CPU times: total: 3min 5s
Wall time: 6min 37s


{'max_depth': 399, 'n_estimators': 9}

In [20]:
models_scores_list['RandomForest'] = [best_params_forest,best_score_forest]

### LGBMClassifier

In [21]:
params_lgbm = {
    'objective': ['binary'],
    'max_depth': range(6,8,1),
    'learning_rate':np.arange(0.021,0.023,0.001),
    'boosting_type': ['gbdt'],
    'metric': ['f1'],
    'verbosity': [-10],
    'random_state': [12345],
    'n_estimators': range(400,501,50),    
    }

model_lgbm = LGBMClassifier(random_state = 12345,class_weight='balanced',n_jobs = -1)

In [22]:
%%time
model_grid_lgbm,best_params_lgbm,best_score_lgbm = grid_search_params(model_lgbm,params_lgbm,
                                                                   tf_idf_train,target_train)

display(HTML(f"<font color='red'>f1 LGBMClassifier:{best_score_lgbm:.3f}</font>"))
best_params_lgbm

Fitting 3 folds for each of 12 candidates, totalling 36 fits


CPU times: total: 7min 44s
Wall time: 16min 45s


{'boosting_type': 'gbdt',
 'learning_rate': 0.022000000000000002,
 'max_depth': 7,
 'metric': 'f1',
 'n_estimators': 500,
 'objective': 'binary',
 'random_state': 12345,
 'verbosity': -10}

In [23]:
models_scores_list['LGBMClassifier'] = [best_params_lgbm,best_score_lgbm]

***Вывод:***<br>

*Ни она модель не прошла минимальный установленный порог метрики f1*<br>
*Попробуем в лучшую по метрике `f1` модель LogisticRegression передать новую матрицу векторизованных методом `CountVectorizer` признаков*

In [24]:
vectorized_train,vectorized_test = vectorization(features_train,features_test,CountVectorizer)

print('Размер матрицы векторизованных признаков для обучающей выборки:',vectorized_train.shape)
print('Размер матрицы векторизованных признаков для тестовой выборки:',vectorized_test.shape)

Размер матрицы векторизованных признаков для обучающей выборки: (127656, 144110)
Размер матрицы векторизованных признаков для тестовой выборки: (31915, 144110)


### LogisticRegression - CountVectorizer

In [25]:
params_lr_CV = {"max_iter":range(68,72,1)}
model_lr_CV = LogisticRegression(random_state = 12345,class_weight='balanced',n_jobs = -1, max_iter=250)

In [26]:
%%time
model_grid_lr_CV,best_params_lr_CV,best_score_lr_CV = grid_search_params(model_lr_CV,
                            params_lr_CV,vectorized_train,target_train)

display(HTML(f"<font color='green'>f1 LogisticRegression: {best_score_lr_CV:.3f}</font>"))
best_params_lr_CV

Fitting 3 folds for each of 4 candidates, totalling 12 fits


CPU times: total: 453 ms
Wall time: 21.5 s


{'max_iter': 70}

In [27]:
models_scores_list['LogisticRegression - CountVectorizer'] = [best_params_lr_CV,best_score_lr_CV]

In [28]:
models_scores = pd.DataFrame(models_scores_list,index=['params', 'f1-score'])
models_scores

LogisticRegression                           RandomForest  \
params     {'max_iter': 65}  {'max_depth': 399, 'n_estimators': 9}   
f1-score           0.746793                               0.627933   

                                                                                                                                                                                 LGBMClassifier  \
params    {'boosting_type': 'gbdt', 'learning_rate': 0.022000000000000002, 'max_depth': 7, 'metric': 'f1', 'n_estimators': 500, 'objective': 'binary', 'random_state': 12345, 'verbosity': -10}   
f1-score                                                                                                                                                                               0.713566   

         LogisticRegression - CountVectorizer  
params                       {'max_iter': 70}  
f1-score                             0.752134

### Вывод

***С поставленной задачей минимальной метрики f1 > 0.75 смогла справиться лишь модель LogisticRegression с методом векторизации CountVectorizer***

### Проверка моделей на тестовой выборке

In [29]:
#Обучим модель заново уже без GridSearch на ранее полученных, лучших параметрах.

model_log_best = LogisticRegression(random_state = 12345,class_weight='balanced',n_jobs = -1, max_iter=250)
model_log_best.fit(vectorized_train,target_train)

#получим предсказания модели для тестовой выборки, чтобы сравнить имеющимся вектором целевого признака.
predicted = model_log_best.predict(vectorized_test)

#Вычислим f1 меру для полученных предсказаний.
score_best = f1_score(target_test,predicted)
display(HTML(f"<font color='blue'>f1_score лучшей модели LinearRegression с векторизацией CountVectorizer:\
{score_best:.3f}</font>"))

<div class="alert alert-success">
<b>✔️ Комментарий ревьювера:</b> Поздравляю, необходимый порог по значению F-меры пройден!
</div>

## Выводы

***Среди исследованных моделей лучшей и удовлетворяющей условию задачи по минимальному порогу `f1`меры (среднего гармонического полноты-`recall`и точности-`precision`) стала модель логистической регрессии, обученная на выборке признаков, полученных векторизатором CountVectorizer.*** <br>
***При проверке на тестовой выборке модель показала `f1`= 0.756***

<font color=blue>
<b> ✔️ 👍Заключение ревьювера:</b> Михаил, было приятно проверять твою работу, она выполнена на хорошем уровне! 
   
Код структурирован, здорово, что попробовал разные модели для обучения и сделал перебор параметров.
    
Особых замечаний по работе нет, я отметил только рекомендации.
    
Отправляю работу для ознакомления с ходом проверки, а также если есть вопросы — смело задавай их, я постараюсь тебе помочь 😊

Жду твоего ответа!
 </font>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны